rust/src/lib.rs

```diff
@@ -2679,6 +2792,105 @@ pub unsafe extern "C" fn zcashlc_shield_funds(
+        let mut db_data = unsafe { wallet_db(db_data, db_data_len, network)? };
+
+        let usk = unsafe { decode_usk(usk_ptr, usk_len) }?;
+
+        let memo_bytes = if memo.is_null() {
+            MemoBytes::empty()
+        } else {
+            MemoBytes::from_bytes(unsafe { slice::from_raw_parts(memo, 512) })
+                .map_err(|e| anyhow!("Invalid MemoBytes: {}", e))?
+        };
+
+        let shielding_threshold = NonNegativeAmount::from_u64(shielding_threshold)
+            .map_err(|()| anyhow!("Invalid amount, out of range"))?;
+
+        let account = db_data
+            .get_account_for_ufvk(&usk.to_unified_full_viewing_key())?
+            .ok_or_else(|| anyhow!("Spending key not recognized."))?;
+
+        let taddrs: Vec<TransparentAddress> = db_data
+            .get_target_and_anchor_heights(NonZeroU32::MIN)
+            .map_err(|e| anyhow!("Error while fetching anchor height: {}", e))
+            .and_then(|opt_anchor| {
+                opt_anchor
+                    .map(|(target, _)| target) // Include unconfirmed funds.
+                    .ok_or_else(|| anyhow!("height not available; scan required."))
+            })
+            .and_then(|anchor| {
+                db_data
+                    .get_transparent_balances(account, anchor)
+                    .map_err(|e| {
+                        anyhow!(
+                            "Error while fetching transparent balances for {:?}: {}",
+                            account,
+                            e,
+                        )
+                    })
+            })?
+            .keys()
+            .cloned()
+            .collect();
+
+        let fee_rule = if use_zip317_fees {
+            StandardFeeRule::Zip317
+        } else {
+            #[allow(deprecated)]
+            StandardFeeRule::PreZip313
+        };
+
+        let input_selector = GreedyInputSelector::new(
+            standard::SingleOutputChangeStrategy::new(fee_rule, Some(memo_bytes)),
+            DustOutputPolicy::default(),
+        );
+
+        let proposal = propose_shielding(
+            &mut db_data,
+            &network,
+            &input_selector,
+            shielding_threshold,
+            &taddrs,
+            min_confirmations,
+        )
+        .map_err(
+            |e: Error<
+                SqliteClientError,
+                Infallible,
+                GreedyInputSelectorError<FeeError, Infallible>,
+                FeeError,
+            >| anyhow!("Error while proposing transfer: {}", e),
+        )?;
+
+        if let Some(proto) = proposal::Proposal::from_standard_proposal(&network, &proposal) {
+            let mut buf = Vec::new();
+            buf.reserve(proto.encoded_len());
+            proto.encode(&mut buf).unwrap();
+            let proposal_bytes = Box::new(buf);
+            unsafe { std::ptr::copy(proposal_bytes.as_ptr(), proposal_bytes_ret, proposal_bytes.len()); }
+            Ok(true)
```

Comment:
This would almost certainly have caused a crash. In this code, `proposal_bytes_ret` is allocated by the Swift caller, but transaction proposal encodings are variable length (and protobuf encodings are non-canonical so can even vary in length with re-encoding), and there is no way for the caller to know what length they need to allocate.

In #111 I address this by allocating the memory on the Rust side.